pkg/pull/pull.go

```diff
@@ -293,6 +299,27 @@ func Pull(upstreamURI string, pullOptions PullOptions) (string, error) {
 	}
 	log.FinishSpinner()
 
+	kotsKinds, err := kotsutil.LoadKotsKindsFromPath(u.GetUpstreamDir(writeUpstreamOptions))
+	if err != nil {
+		return "", errors.Wrap(err, "failed to load kotskinds")
+	}
+
+	registrySettings := registrytypes.RegistrySettings{
+		Hostname:   pullOptions.RewriteImageOptions.Host,
+		Namespace:  pullOptions.RewriteImageOptions.Namespace,
+		Username:   pullOptions.RewriteImageOptions.Username,
+		Password:   pullOptions.RewriteImageOptions.Password,
+		IsReadOnly: pullOptions.RewriteImageOptions.IsReadOnly,
+	}
+
+	needsConfig, err := kotsadmconfig.NeedsConfiguration(pullOptions.AppSlug, pullOptions.AppSequence, pullOptions.AirgapRoot != "", kotsKinds, registrySettings)
+	if err != nil {
+		return "", errors.Wrap(err, "failed to check if version needs configuration")
+	}
+	if needsConfig {
+		return "", ErrConfigNeeded
+	}
```

Comment:
This causes the `Pull(...)` function to return if `needsConfig` and we do not write midstreams (which is expected since we aren't rendering the version).  Unfortunately, this seems to also result in the airgap images not getting pushed since that is part of `writeMidstream(...)`: https://github.com/replicatedhq/kots/blob/main/pkg/pull/pull.go#L579.

Reply:
That makes sense.  Nice catch.